### PR TITLE
respect PodCIDR of node

### LINKED ISSUE #17: [kwok] Respect PodCIDR of Node
/kind feature

### PR DESCRIPTION
**Pull request by carlory**
Fixes #17 

/kind feature

### CODE REVIEW

**Comment by wzshiming** [comment]
> @wzshiming how to join the org? I'm a kubernetes member. [kubernetes/org#2402](https://github.com/kubernetes/org/issues/2402)
> 
> Can you give me some guide?

If you are already a member of kubernetes, just submit an issue to kubernetes/org to apply for membership in kubernetes-sigs without any additional requirements

File: pkg/kwok/controllers/pod_controller.go [hunk 1]
```diff
@@ -363,6 +368,21 @@ func (c *PodController) configurePod(pod *corev1.Pod) ([]byte, error) {
...
+		c.ipPool.Use(pod.Status.PodIP)
+	}
+
+	nodeInfo := c.nodeInfoGetFunc(pod.Spec.NodeName)
+	if nodeInfo == nil || nodeInfo.CidrIPNet == nil || nodeInfo.IPPool == nil {
+		return
+	}
+
+	if nodeInfo.CidrIPNet.Contains(net.ParseIP(pod.Status.PodIP)) {
+		nodeInfo.IPPool.Use(pod.Status.PodIP)
```
Comment (wzshiming) [inline pkg/kwok/controllers/pod_controller.go:382]: Seems to have missed the logic for reclaiming Pod IP

File: pkg/kwok/controllers/pod_controller_test.go [hunk 2]
```diff
@@ -49,6 +50,22 @@ func TestPodController(t *testing.T) {
 				NodeName: "node0",
 			},
 		},
+		&corev1.Pod{
```
Comment (wzshiming) [inline pkg/kwok/controllers/pod_controller_test.go:53]: It is better to test the complete allocation and recycling behavior of Node's PodCIDR and default CIDR in a new test function

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: pkg/kwok/controllers/pod_controller.go (hunk 1 above)
Comment (carlory) [reply pkg/kwok/controllers/pod_controller.go:378]: @wzshiming fixed

File: pkg/kwok/controllers/pod_controller.go [hunk 3]
```diff
@@ -305,12 +312,8 @@ func (c *PodController) WatchPods(ctx context.Context, lockChan, deleteChan chan
...
 				case watch.Deleted:
 					pod := event.Object.(*corev1.Pod)
-					if c.nodeHasFunc(pod.Spec.NodeName) {
-						// Recycling PodIP
-						if pod.Status.PodIP != "" && c.cidrIPNet.Contains(net.ParseIP(pod.Status.PodIP)) {
-							c.ipPool.Put(pod.Status.PodIP)
-						}
-					}
+					// Recycling PodIP
+					c.reclaimPodIP(pod)
```
Comment (wzshiming) [inline pkg/kwok/controllers/pod_controller.go:316]:
```suggestion
				if c.nodeHasFunc(pod.Spec.NodeName) {
					// Recycling PodIP
					c.reclaimPodIP(pod)
				}
```

File: pkg/kwok/controllers/utils.go [hunk 4]
```diff
@@ -160,43 +168,71 @@ func (p *parallelTasks) Wait() {
 	p.wg.Wait()
 }
 
-type stringSets struct {
+// nodeInfo holds information about a node
+type nodeInfo struct {
```
Comment (wzshiming) [inline pkg/kwok/controllers/utils.go:172]:
```suggestion
type NodeInfo struct {
```

**Review by wzshiming — CHANGES_REQUESTED** [review]

**Pull request closed without merging**